### ISSUE TITLE
SimpleFunctionRegistry cannot convert generic type messages when using Functional message listeners

### ISSUE THREAD

**Issue by CH-jakegale** [spring-cloud/spring-cloud-function#536]
This is implicitly in Spring Integration, so I'm not sure if this is the right place. If it is not, please direct me and I can post there. However, the issue lies when using Spring Cloud 3.0+.

When using a Function bean as the message processor that takes a generic type as the input type `Message<C>` where C extends some class, we cannot marshal the message into the required type.

**To Reproduce**
I have a Function `@Bean` to act as a message listener, however this takes a generic type message, 
e.g.
```
public Clazz <C extends SomeClass> {

@Bean
Function<Message<C> , Message<C>> process() {
    return message -> doProcess(message);
}
```

When processing a message from the RabbitMQ binder (though I assume this is not unique to this binder implementation) the message is passed through, and the function bean is invoked through ` org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry`.

This has the following boolean method:
```
        private boolean messageNeedsConversion(Type rawType, Message<?> message) {
            Boolean skipConversion = message.getHeaders().containsKey("skip-type-conversion") ? (Boolean)message.getHeaders().get("skip-type-conversion", Boolean.class) : false;
            if (skipConversion) {
                return false;
            } else {
                return rawType instanceof Class && !(message.getPayload() instanceof Optional) && !message.getPayload().getClass().isAssignableFrom((Class)rawType);
            }
        }
```
and we want to look specifically at this line
```
 return rawType instanceof Class && !(message.getPayload() instanceof Optional) && !message.getPayload().getClass().isAssignableFrom((Class)rawType);
```
and even more specifically at `rawType instanceof Class`.
rawType in this case is `C`, the generic, and that boolean condition returns false, meaning the returns method returns false and the message will skip conversion.
Since the message is of type `byte[]` this cannot be cast into the resultant `SomeClass`.

**Version of the framework**
- Spring Cloud Stream:  3.0.4.RELEASE
- Spring Cloud Function is 3.0.6.RELEASE
- Spring Integration Core: 5.2.3.RELEASE

**Additional Context** 
The non-functional approach, with legacy `@StreamListener` works fine with the generics.
The `TestBinder` does not fail to convert the incoming messages, though that is likely because I pass the definitive type as an object instead of a `byte[]` from the external binder.
I am getting this error when writing a message from to the queue from the rabbit UI.

**Comment by olegz** [comment]
Actually looking further at it we can't really do much about it, since we don't know the concrete type you are looking for. For example in a typical case where you're passing JSON representing your POJO we need to know the type so we can provide this type to a Message converter so it can use it to create an instance of this type from the json string/byte[]. I actually explain the mechanics of it [here](https://docs.spring.io/spring-cloud-stream/docs/3.1.0.M1/reference/html/spring-cloud-stream.html#_using_streamlistener_for_content_based_routing)
The only way I can see it could work is with java serializatioin (e.g., Kryo converter), where the byte[] coming in is actually a serialized java stream representing the object. Obviously the content type would be different and you would have to register the appropriate message converter (KryoMessageConverter) as we no longer provide one out of the box since passing natively serialized objects is a non-recommended practice that could lead to a slew of other issues. 

In other words there is really nothing for us to do here.
